File: src/main/java/one/oktw/sponge/event/ForceGamemode.java
package one.oktw.sponge.event;

import org.spongepowered.api.data.key.Keys;
import org.spongepowered.api.entity.living.player.Player;
import org.spongepowered.api.event.Listener;
import org.spongepowered.api.event.entity.MoveEntityEvent;
import org.spongepowered.api.event.filter.Getter;
import org.spongepowered.api.world.World;
import org.spongepowered.api.world.storage.WorldProperties;

public class ForceGamemode {
    @Listener
    public void onMoveEntityEventEventTeleport(MoveEntityEvent.Teleport event, @Getter("getTargetEntity") Player player) {
        World from = event.getFromTransform().getExtent();
        World to = event.getToTransform().getExtent();

        WorldProperties properties = to.getProperties();

        if (!from.equals(to)) {
            if (!player.hasPermission("oktw.world.gamemode")) {
                if (!properties.getGameMode().equals(player.gameMode().get())) {
                    player.offer(Keys.GAME_MODE, properties.getGameMode());
                }
            }
        }
    }
}

File: src/main/kotlin/one/oktw/galaxy/event/ForceGamemode.kt
package one.oktw.galaxy.event

import org.spongepowered.api.data.key.Keys
import org.spongepowered.api.entity.living.player.Player
import org.spongepowered.api.entity.living.player.gamemode.GameMode
import org.spongepowered.api.event.Listener
import org.spongepowered.api.event.entity.MoveEntityEvent
import org.spongepowered.api.event.filter.Getter
import org.spongepowered.api.event.network.ClientConnectionEvent

class ForceGamemode {
    @Listener
    fun onMoveEntityEventEventTeleport(event: MoveEntityEvent.Teleport, @Getter("getTargetEntity") player: Player) {
        val from = event.fromTransform.extent
        val to = event.toTransform.extent

        val properties = to.properties

        if (from != to) {
            if (!player.hasPermission("oktw.world.gamemode")) {
                if (properties.gameMode != player.gameMode().get()) {
                    player.offer<GameMode>(Keys.GAME_MODE, properties.gameMode)
                }
            }
        }
    }

    @Listener
    fun onPlayerJoin(event: ClientConnectionEvent.Join, @Getter("getTargetEntity") player: Player) {
        val properties = player.world.properties

        if (!player.hasPermission("oktw.world.gamemode")) {
            if (properties.gameMode != player.gameMode().get()) {
                player.offer<GameMode>(Keys.GAME_MODE, properties.gameMode)
            }
        }
    }
}
